Upgrade caldav to 1.0.1

daniele-athome opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Bump version of caldav library: date_search was deprecated in favour of search which also implements client-side recurring events expansion.

## python-caldav version bump

* Changelog relevant to the new search method: https://github.com/python-caldav/caldav/blob/master/changelog-0.11.md
* Changelog of latest version (mainly fixes compared to 0.11): https://github.com/python-caldav/caldav/blob/master/changelog-1.0.md
* Changes: https://github.com/python-caldav/caldav/compare/v0.9.1...v1.0.1

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [x] Dependency upgrade
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #40127
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [ ] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

daniele-athome commented:
Ok I'll admit I didn't run the full test suite locally, but I didn't think a version bump would cause this. I'll need to investigate. Bringing the PR back to draft.

tobixen commented on homeassistant/components/caldav/calendar.py:
```diff
@@ -172,7 +173,13 @@ async def async_get_events(
+                self.calendar.search,
+                start=start_date,
+                end=end_date,
+                comp_class=caldav.Event,
```
`comp_class=caldav.Event` may be replaced with `event=True`

daniele-athome commented:
I've got unrelated errors in local tests, so I'm marking this as ready so the CI tests will hopefully provide more insights.

allenporter commented:
Approving to run CI

daniele-athome commented:
```
  File "/home/daniele/Development/home-assistant/homeassistant/components/caldav/calendar.py", line 236, in update
    new_event = event.copy()
  File "/home/daniele/Development/home-assistant/.venv/lib/python3.10/site-packages/caldav/objects.py", line 1909, in copy
    obj.url = obj.generate_url()
  File "/home/daniele/Development/home-assistant/.venv/lib/python3.10/site-packages/caldav/objects.py", line 2001, in generate_url
    return self.parent.url.join(quote(self.id.replace("/", "%2F")) + ".ics")
```

I think I need to bring this to upstream.

daniele-athome commented:
Passing `keep_uid=True` to `event.copy` (components/caldav/calendar.py:236) makes it work, so it's the recurring events. I think it has something to do with how we expand recurring events, [stuff we were discussing over here](https://github.com/python-caldav/caldav/pull/222#issuecomment-1295880929), @tobixen any thoughs?

I'll continue tomorrow.

tobixen commented:
> ```
>   File "/home/daniele/Development/home-assistant/homeassistant/components/caldav/calendar.py", line 236, in update
>     new_event = event.copy()
>   File "/home/daniele/Development/home-assistant/.venv/lib/python3.10/site-packages/caldav/objects.py", line 1909, in copy
>     obj.url = obj.generate_url()
>   File "/home/daniele/Development/home-assistant/.venv/lib/python3.10/site-packages/caldav/objects.py", line 2001, in generate_url
>     return self.parent.url.join(quote(self.id.replace("/", "%2F")) + ".ics")
> ```

I see a traceback, but no errors here ... what is the actual error message?

allenporter commented:
Here is an example: https://github.com/home-assistant/core/actions/runs/3920550422/jobs/6702347492#step:10:8387
```
Traceback (most recent call last):
  File "/home/runner/work/core/core/homeassistant/helpers/entity_platform.py", line 507, in _async_add_entity
    await entity.async_device_update(warning=False)
  File "/home/runner/work/core/core/homeassistant/helpers/entity.py", line 739, in async_device_update
    await task
  File "/opt/hostedtoolcache/Python/3.9.16/x64/lib/python3.9/concurrent/futures/thread.py", line 58, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/home/runner/work/core/core/homeassistant/components/caldav/calendar.py", line 147, in update
    self.data.update()
  File "/home/runner/work/core/core/homeassistant/util/__init__.py", line 190, in wrapper
    result = method(*args, **kwargs)
  File "/home/runner/work/core/core/homeassistant/components/caldav/calendar.py", line 236, in update
    new_event = event.copy()
  File "/home/runner/work/core/core/venv/lib/python3.9/site-packages/caldav/objects.py", line 1909, in copy
    obj.url = obj.generate_url()
  File "/home/runner/work/core/core/venv/lib/python3.9/site-packages/caldav/objects.py", line 2001, in generate_url
    return self.parent.url.join(quote(self.id.replace("/", "%2F")) + ".ics")
AttributeError: 'NoneType' object has no attribute 'url'
```

@daniele-athome i think it's much easier to get these locally with `pytest tests/components/caldav/` fwiw

tobixen commented:
The copy method looks like this ...

```
   def copy(self, keep_uid=False, new_parent=None):
        """
        Events, todos etc can be copied within the same calendar, to another
        calendar or even to another caldav server
        """
        obj = self.__class__(
            parent=new_parent or self.parent,
            data=self.data,
            id=self.id if keep_uid else str(uuid.uuid1()),
        )
        if new_parent or not keep_uid:
            obj.url = obj.generate_url()
        else:
            obj.url = self.url
        return obj
```

... so yes, if keep_uid is set to True, the event will keep it's old URL and the generate_url logic will not run.  But parent (the calendar, that is) should be set in any case, it shouldn't be `None`.

I will check up some few things on my side and check that if code section is sufficiently covered with tests or not ... but first I need to sleep ...

tobixen commented:
Can you try to run the tests with the issue_250 branch of the caldav project?

daniele-athome commented:
> I see a traceback, but no errors here ... what is the actual error message?

Sorry! Paste error.

> @daniele-athome i think it's much easier to get these locally with pytest tests/components/caldav/ fwiw

Yes thank you, I realized that some time later.

> Can you try to run the tests with the issue_250 branch of the caldav project?

All tests fail on this assertion: https://github.com/python-caldav/caldav/blob/issue_250/caldav/objects.py#L1693

tobixen commented:
> All tests fail on this assertion: https://github.com/python-caldav/caldav/blob/issue_250/caldav/objects.py#L1693

Good!  Do you have a traceback?

daniele-athome commented:
Sorry I didn't think it was useful. Here are some of the failed tests:

```
――――――――――――――――――――――――――――――――――――――――――――――――――― ERROR at setup of test_event_rrule_hourly_off_last ―――――――――――――――――――――――――――――――――――――――――――――――――――

    @pytest.fixture(name="calendar")
    def mock_private_cal():
        """Mock a private calendar."""
>       _calendar = _mock_calendar("Private")

tests/components/caldav/test_calendar.py:311: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
tests/components/caldav/test_calendar.py:356: in _mock_calendar
    events.append(Event(None, "%d.ics" % idx, event, None, str(idx)))
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = Event(Event: 0.ics), client = None, url = '0.ics'
data = 'BEGIN:VCALENDAR\nVERSION:2.0\nPRODID:-//E-Corp.//CalDAV Client//EN\nBEGIN:VEVENT\nUID:1\nDTSTAMP:20171125T000000Z\nDT...180000Z\nSUMMARY:This is a normal event\nLOCATION:Hamburg\nDESCRIPTION:Surprisingly rainy\nEND:VEVENT\nEND:VCALENDAR\n'
parent = None, id = '0', props = None

    def __init__(
        self, client=None, url=None, data=None, parent=None, id=None, props=None
    ):
        """
        CalendarObjectResource has an additional parameter for its constructor:
         * data = "...", vCal data for the event
        """
        super(CalendarObjectResource, self).__init__(
            client=client, url=url, parent=parent, id=id, props=props
        )
>       assert self.parent
E       AssertionError

../python-caldav/caldav/objects.py:1693: AssertionError
----------------------------------------------------------------- Captured stderr setup ------------------------------------------------------------------
DEBUG:asyncio:Using selector: EpollSelector
------------------------------------------------------------------- Captured log setup -------------------------------------------------------------------
2023-01-16 07:09:02.542 DEBUG    MainThread asyncio:selector_events.py:54 Using selector: EpollSelector
                                                                                                                                            93% █████████▍

―――――――――――――――――――――――――――――――――――――――――――――――――――― ERROR at setup of test_event_rrule_hourly_ended ―――――――――――――――――――――――――――――――――――――――――――――――――――――

    @pytest.fixture(name="calendar")
    def mock_private_cal():
        """Mock a private calendar."""
>       _calendar = _mock_calendar("Private")

tests/components/caldav/test_calendar.py:311: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
tests/components/caldav/test_calendar.py:356: in _mock_calendar
    events.append(Event(None, "%d.ics" % idx, event, None, str(idx)))
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = Event(Event: 0.ics), client = None, url = '0.ics'
data = 'BEGIN:VCALENDAR\nVERSION:2.0\nPRODID:-//E-Corp.//CalDAV Client//EN\nBEGIN:VEVENT\nUID:1\nDTSTAMP:20171125T000000Z\nDT...180000Z\nSUMMARY:This is a normal event\nLOCATION:Hamburg\nDESCRIPTION:Surprisingly rainy\nEND:VEVENT\nEND:VCALENDAR\n'
parent = None, id = '0', props = None

    def __init__(
        self, client=None, url=None, data=None, parent=None, id=None, props=None
    ):
        """
        CalendarObjectResource has an additional parameter for its constructor:
         * data = "...", vCal data for the event
        """
        super(CalendarObjectResource, self).__init__(
            client=client, url=url, parent=parent, id=id, props=props
        )
>       assert self.parent
E       AssertionError

../python-caldav/caldav/objects.py:1693: AssertionError
----------------------------------------------------------------- Captured stderr setup ------------------------------------------------------------------
DEBUG:asyncio:Using selector: EpollSelector
------------------------------------------------------------------- Captured log setup -------------------------------------------------------------------
2023-01-16 07:09:02.628 DEBUG    MainThread asyncio:selector_events.py:54 Using selector: EpollSelector
                                                                                                                                            95% █████████▌

――――――――――――――――――――――――――――――――――――――――――――――――――――――――――― ERROR at setup of test_get_events ――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――

    @pytest.fixture(name="calendar")
    def mock_private_cal():
        """Mock a private calendar."""
>       _calendar = _mock_calendar("Private")

tests/components/caldav/test_calendar.py:311: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
tests/components/caldav/test_calendar.py:356: in _mock_calendar
    events.append(Event(None, "%d.ics" % idx, event, None, str(idx)))
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = Event(Event: 0.ics), client = None, url = '0.ics'
data = 'BEGIN:VCALENDAR\nVERSION:2.0\nPRODID:-//E-Corp.//CalDAV Client//EN\nBEGIN:VEVENT\nUID:1\nDTSTAMP:20171125T000000Z\nDT...180000Z\nSUMMARY:This is a normal event\nLOCATION:Hamburg\nDESCRIPTION:Surprisingly rainy\nEND:VEVENT\nEND:VCALENDAR\n'
parent = None, id = '0', props = None

    def __init__(
        self, client=None, url=None, data=None, parent=None, id=None, props=None
    ):
        """
        CalendarObjectResource has an additional parameter for its constructor:
         * data = "...", vCal data for the event
        """
        super(CalendarObjectResource, self).__init__(
            client=client, url=url, parent=parent, id=id, props=props
        )
>       assert self.parent
E       AssertionError

../python-caldav/caldav/objects.py:1693: AssertionError
----------------------------------------------------------------- Captured stderr setup ------------------------------------------------------------------
DEBUG:asyncio:Using selector: EpollSelector
------------------------------------------------------------------- Captured log setup -------------------------------------------------------------------
2023-01-16 07:09:02.715 DEBUG    MainThread asyncio:selector_events.py:54 Using selector: EpollSelector
                                                                                                                                            98% █████████▊

――――――――――――――――――――――――――――――――――――――――――――――――――― ERROR at setup of test_get_events_custom_calendars ―――――――――――――――――――――――――――――――――――――――――――――――――――

    @pytest.fixture(name="calendar")
    def mock_private_cal():
        """Mock a private calendar."""
>       _calendar = _mock_calendar("Private")

tests/components/caldav/test_calendar.py:311: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
tests/components/caldav/test_calendar.py:356: in _mock_calendar
    events.append(Event(None, "%d.ics" % idx, event, None, str(idx)))
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = Event(Event: 0.ics), client = None, url = '0.ics'
data = 'BEGIN:VCALENDAR\nVERSION:2.0\nPRODID:-//E-Corp.//CalDAV Client//EN\nBEGIN:VEVENT\nUID:1\nDTSTAMP:20171125T000000Z\nDT...180000Z\nSUMMARY:This is a normal event\nLOCATION:Hamburg\nDESCRIPTION:Surprisingly rainy\nEND:VEVENT\nEND:VCALENDAR\n'
parent = None, id = '0', props = None

    def __init__(
        self, client=None, url=None, data=None, parent=None, id=None, props=None
    ):
        """
        CalendarObjectResource has an additional parameter for its constructor:
         * data = "...", vCal data for the event
        """
        super(CalendarObjectResource, self).__init__(
            client=client, url=url, parent=parent, id=id, props=props
        )
>       assert self.parent
E       AssertionError

../python-caldav/caldav/objects.py:1693: AssertionError
```

I could paste them all but they really are pretty much identical.

tobixen commented:
> self = Event(Event: 0.ics), client = None, url = '0.ics'

This is from the tests in the homeassistant code?  The `Event` is created without a parent, and this causes problems in the `copy`-method.  A calendar object has to be passed in the constructor.

tobixen commented:
Since the line creating a caldav `Event` without a parent calendar seems to come from external test code, I deem it's not a problem with the caldav library per se.

I've now released 1.0.0 and I recommend to use that one rather than 0.11.

[Release notes](https://github.com/python-caldav/caldav/blob/master/changelog-1.0.md)

daniele-athome committed:
Upgrade caldav (fix #40127)

date_search was deprecated in favour of search which also implements client-side recurring events expansion

daniele-athome commented:
Ok tests are passing now. I've passed the mock calendar to the Event constructor as parent.

allenporter commented:
Can you update the PR description to include a link to the caldav changelog? Thank you

daniele-athome commented:
> Can you update the PR description to include a link to the caldav changelog? Thank you

Done, both versions.

allenporter commented:
I've updated the PR description to link to the changes between the old version and new version: https://github.com/python-caldav/caldav/compare/v0.9.1...v1.0.1
